mlflow/utils/file_utils.py

```diff
@@ -588,6 +594,76 @@ def download_file_using_http_uri(http_uri, download_path, chunk_size=100000000,
                 output_file.write(chunk)
 
 
+def download_chunk(request_index, chunk_size, headers, download_path, http_uri):
+    range_start = chunk_size * request_index
+    range_end = str(range_start + chunk_size - 1)
+    combined_headers = {**headers, "Range": f"bytes={range_start}-{range_end}"}
+    with cloud_storage_http_request(
+        "get", http_uri, stream=False, headers=combined_headers
+    ) as response:
+        with open(download_path, "r+b") as f:
+            f.seek(range_start)
+            f.write(response.content)
+
+
+def parallelized_download_file_using_http_uri(
+    http_uri, download_path, file_size, uri_type, chunk_size, headers=None
+):
+    """
+    Downloads a file specified using the `http_uri` to a local `download_path`. This function
+    sends multiple requests in parallel each specifying its own desired byte range as a header,
+    then reconstructs the file from the downloaded chunks. This allows for downloads of large files
+    without OOM risk.
+
+    Note : This function is meant to download files using presigned urls from various cloud
+            providers.
+    Returns a dict of chunk index : exception, if one was thrown for that index.
+    """
+    num_requests = int(math.ceil(file_size / float(chunk_size)))
+    # Create file if it doesn't exist. We should do this here before sending to the workers
+    # so they can each individually seek to their respective positions and write chunks
+    # without overwriting.
+    f = open(download_path, "wb")
```

Comment:
Is this file cleaned up if the download fails?

Reply:
boto3 downloads the target object into a temporary file and renames the file:

https://github.com/boto/s3transfer/blob/f225103e69f2e2b61e8933ad816519229afaf088/s3transfer/download.py#L168

Maybe we could do the same thing.